### PR TITLE
Add BindingExpression attribute which lets you specify a parameter th…

### PR DESCRIPTION
**Pull request by MikeStall**
…at resolves against binding data.

SDK implicitly would do this for C# parameters without attributes, but:
1. the explicit attribute lets you do this for other languages and specify in Function.json, Script.
2. the SDK default still only lets x bind to {x}.
Fixes https://github.com/Azure/azure-webjobs-sdk/issues/1498
This already existed in test, moving to product.

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs.Host/Bindings/BindingExpressionAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+using Microsoft.Azure.WebJobs.Description;
+using Microsoft.Azure.WebJobs.Host.Config;
+
+namespace Microsoft.Azure.WebJobs
+{
+    /// <summary>
+    /// Attribute to allow binding directly to binding data. 
+    /// </summary>
+    [Binding]
+    public class BindingExpressionAttribute : Attribute
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/BindingExpressionAttribute.cs]: I think this attribute should be defined in the core project (peer to QueueAttribute, etc.)

File: src/Microsoft.Azure.WebJobs.Host/Bindings/BindingExpressionAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Azure.WebJobs.Description;
+using Microsoft.Azure.WebJobs.Host.Config;
+
+namespace Microsoft.Azure.WebJobs
+{
+    /// <summary>
+    /// Attribute to allow binding directly to binding data.
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/BindingExpressionAttribute.cs]: Can you put a more user friendly/descriptive help text here? Something that would make people understand how to use this if they see it in intellisense :)

File: src/Microsoft.Azure.WebJobs.Host/Bindings/BindingExpressionAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        /// Constructor
+        /// </summary>
+        /// <param name="source">binding expression, such as '{x}' or '%appsetting%'. </param>
+        public BindingExpressionAttribute(string source)
+        {
+            this.Source = source;
+        }
+
+        /// <summary>
+        /// Binding expression for where to pull data.
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/BindingExpressionAttribute.cs]: Improve xml doc

File: src/Microsoft.Azure.WebJobs.Host/Bindings/BindingExpressionExtensionConfigProvider.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+namespace Microsoft.Azure.WebJobs
+{
+    /// <summary>
+    /// Extension provider for <see cref="BindingExpressionAttribute"/>
+    /// </summary>
+    internal class BindingExpressionExtensionConfigProvider : IExtensionConfigProvider
+    {
+        public void Initialize(ExtensionConfigContext context)
+        {
+            var rule = context.AddBindingRule<BindingExpressionAttribute>();
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Bindings/BindingExpressionExtensionConfigProvider.cs]: I don't think these input bindings will be immediately accessible to script functions without changes to the runtime. Have you tried this end to end yet?

**Comment by fabiocav** [comment]
Are we still planning on working on this? Or should this be closed?

**Comment by mathewc** [comment]
Since the PR still has unaddressed comments and we don't have any feature drivers requiring this currently, I vote we don't add this now - it can always be added later in a non-breaking way.

**Comment by fabiocav** [comment]
Sounds good. Closing. We can re-open if needed.

**Pull request closed without merging**